genesis/utils/path_planing.py

```diff
@@ -0,0 +1,995 @@
+from abc import ABC, abstractmethod
+import taichi as ti
+import torch
+import torch.nn.functional as F
+import time
+import genesis as gs
+import genesis.utils.geom as gg
+
+
+class PathPlanner(ABC):
+    def __init__(self, entity):
+        self._entity = entity
+        self._solver = entity._solver
+
+        for joint in entity.joints:
+            if joint.type == gs.JOINT_TYPE.FREE:
+                gs.raise_exception("planning for the gs.JOINT_TYPE.FREE is not supported (yet)")
+            elif joint.type == gs.JOINT_TYPE.SPHERICAL:
+                gs.raise_exception("planning for the gs.JOINT_TYPE.SPHERICAL is not supported (yet)")
+
+    @abstractmethod
+    def plan(
+        self,
+        qpos_goal,
+        qpos_start=None,
+    ): ...
+
+    def get_link_transformation(self, robot_g_link_idx, obj_g_link_idx):
+        """
+        get transformation from robot link to the object link
+        robot_g_link_idx: int
+            global link idx of the link of the robot
+        obj_g_link_idx: int
+            global link idx of the base link of the object
+        """
+        robot_T = gg.trans_quat_to_T(
+            self._solver.get_links_pos(links_idx=robot_g_link_idx),
+            self._solver.get_links_quat(links_idx=robot_g_link_idx),
+        )
+        obj_T = gg.trans_quat_to_T(
+            self._solver.get_links_pos(links_idx=obj_g_link_idx), self._solver.get_links_quat(links_idx=obj_g_link_idx)
+        )
+        T_robot_object = torch.linalg.inv(robot_T) @ obj_T
+        return T_robot_object
+
+    def update_object(self, ee_link_idx, obj_link_idx, T_robot_object, envs_idx):
+        if self._solver.n_envs > 0:
+            robot_T = gg.trans_quat_to_T(
+                self._solver.get_links_pos(ee_link_idx, envs_idx=envs_idx),
+                self._solver.get_links_quat(ee_link_idx, envs_idx=envs_idx),
+            )
+        else:
+            robot_T = gg.trans_quat_to_T(
+                self._solver.get_links_pos(ee_link_idx), self._solver.get_links_quat(ee_link_idx)
+            )
+        obj_T = robot_T @ T_robot_object
+        trans, quat = gg.T_to_trans_quat(obj_T)
+        if self._solver.n_envs > 0:
+            self._solver.set_base_links_pos(trans, obj_link_idx, envs_idx=envs_idx)
+            self._solver.set_base_links_quat(quat, obj_link_idx, envs_idx=envs_idx)
+        else:
+            self._solver.set_base_links_pos(trans, obj_link_idx)
+            self._solver.set_base_links_quat(quat, obj_link_idx)
+
+    # ------------------------------------------------------------------------------------
+    # ------------------------------ util funcs ------------------------------------------
+    # ------------------------------------------------------------------------------------
+    def format_input(self, qpos_cur, qpos_goal, qpos_start, envs_idx):
```

Comment:
As I said, it should rather be called `_sanitize_[...]` for consistency with the whole codebase.